Add new shield X-NUCLEO-53L0A1

titouanc opened the pull request:
* Add support for multiple sensors and reconfiguring their I2C address in the vl53l0x driver
* Add the new shield X-NUCLEO-53L0A1 with 3 sensors on the same bus and 4x7 segments display via 2 GPIO expanders
* Add a sample application for the shield

titouanc committed:
boards: shields: x_nucleo_53l0a1: new shield

Add support for the X-NUCLEO-53L0A1 expansion board and an
accompanying sample application

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

erwango reviewed (commented):
Nice work!

I've not been able to test the sample, but at least compiles, and will be compiled by twister, so that's a good point.

Regarding the driver, I've been able to see that it's still functional.
Though, I didn't review in deep: it would have been nice to split the
changes to ease the review.

Can you consider to split a bit this first part ?

titouanc committed:
drivers: sensor: vl53l0x: add driver support for multiple sensors

Up until now, the vl53l0x driver only supported a single device, ie.
the first entry st,vl53l0x in the device tree. To be able to use
multiple sensors at the same time, create one driver data instance per
vl53l0x node in the device tree. Also split the constant driver config
from the runtime data.

Because the vl53l0x address is only configurable with an I2C command,
and is not persisted if the sensor is rebooted, multiple sensors can
be handled only if either:
- They are on different I2C buses
- Their addresses are coonfigured (by some external code) before
  vl53l0x_init is called

Also use the i2c_dt_spec and gpio_dt_spec APIs, as it makes
the code more concise and readable.

Finally, to distinguish the logs mesages from different sensors,
prefix the text with the sensor name.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
drivers: sensor: vl53l0x: run uncrustify

Apply uncrustify on the driver source code before going further

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
drivers: sensor: vl53l0x: add support to reprogram I2C address

The I2C address of the VL53L0X distance sensor can only be programmed
over the I2C bus. To do this:

1. The sensor is powered off or in standby mode
2. Power up the sensor, it boots with a default I2C address (0x29)
3. The I2C master sends a configuration command to set the new address
4. The sensor now communicates at the new address

In case there are more than one such sensor on the bus, they all have
the same address when starting up. We therefore need to first apply
step 1. on all of them. Then, sensor by sensor apply steps 2. to 4.

Because simple designs may not need to reprogram the address, we
introduce a new configuration option CONFIG_VL53L0X_RECONFIGURE_ADDRESS

If this setting is disabled, then the driver behaves as before.

If CONFIG_VL53L0X_RECONFIGURE_ADDRESS is enabled, then the driver does
the following:

- In vl53l0x_init(), apply step 1. This is done when the driver is
  brought up when starting the system
- in vl53l0x_start(), apply steps 2. to 4. This is done when fetching
  a sample, if the sensor has not been started yet.

Also, as cosmetic changes:
- add parenthesis around sub conditions in call to __ASSERT_NO_MSG
- gracefully handle unknown sensor channels in vl53l0x_channel_get

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
drivers: sensor: vl53l0x: wait only 2ms for tboot

As per the VL53L0X datasheet, in 2.9.1 "Power up and boot sequence",
time to boot is 1.2ms max, so we only have to wait at most 2ms.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc commented:
Thank you @erwango for your first remarks.

I addressed your 2 comments above, and I broke up the first patch into smaller ones, which should improve the readability. Please let me know if you need further dissection.

erwango reviewed (changes requested):
Thanks for splitting the driver changes.
Minor changes requested and one question.

titouanc commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -168,10 +168,10 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
 	if (ret) {
 		LOG_ERR(
-		"VL53L0X_SetMeasurementTimingBudgetMicroSeconds failed");
+			"VL53L0X_SetMeasurementTimingBudgetMicroSeconds failed");
```
Yes indeed, though I did this in a subsequent commit https://github.com/zephyrproject-rtos/zephyr/pull/38315/commits/88ae64ece82bdff4083cb0fbbf9a4a75ac5b981a#diff-d58092bb6d79c1da0579e21fb5d1eed6322aa15905562ba99c8e7d0927be914bR183-R187

The rationale is: the first commit of this PR is entirely automated, the rest is written by a human

titouanc commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -208,34 +158,40 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
-	drv_data->vl53l0x.I2cDevAddr = drv_data->dt_addr;
+#ifdef CONFIG_VL53L0X_RECONFIGURE_ADDRESS
+	if (drv_data->dt_addr != VL53L0X_INITIAL_ADDR) {
+		ret = VL53L0X_SetDeviceAddress(&drv_data->vl53l0x, 2 * drv_data->dt_addr);
```
Not entirely sure where this division comes from, as it has been imported like this in https://github.com/zephyrproject-rtos/hal_st/commit/7fa2cac9647643efb3128b9d75561d69f8c2a20c

My best guess is:
* In 7bits addressing, the I2C transaction mode (read/write) is given in the lowest bit of the address, the 7 upper bits being the device address (see https://en.wikipedia.org/wiki/I%C2%B2C#7-bit_addressing)
* From various examples I found across the internet, the ST HAL seems to always store the 7bits address in the upper bits, leaving the lowest one free to be marked as read or write.
* When programming the I2C address though, the device is being given its "true" 7b address, and not the 7+1 bits.
* On the other hand, the Zephyr API always use the "true" I2C address, and will handle the left shift (`<< 1` or `* 2`) when preparing the transaction; so all the functions that communicate with the sensor, ultimately resorting to `i2c_read/write`, have been patched in here.

As a conclusion, this looks like a design difference between I2C subsystem in Zephyr  and the ST HAL. Maybe what we could do would be to remove the division in https://github.com/zephyrproject-rtos/hal_st/blob/master/sensor/vl53l0x/api/core/src/vl53l0x_api.c#L360-L361, make a new release, then update and merge this PR. This could work because the function `VL53L0X_SetDeviceAddress` is not used yet in `main`, but it could potentially break user code as some project might rely on this API (not sure if possible though).

All in all, I'm not sure if this is a better solution than simply multiplying by 2 here :shrug:

erwango replied:
ok, thanks for explanation. Somehow I missed that it was already there.
Let's keep it that way.

titouanc commented on boards/shields/x_nucleo_53l0a1/Kconfig.defconfig:
```diff
@@ -0,0 +1,24 @@
+if SHIELD_X_NUCLEO_53L0A1
+
+config GPIO
+	default y
```
@erwango after removing the above, this fails in buildkite: https://buildkite.com/zephyr/zephyr/builds/40215#43279ba5-fbc3-45aa-a3d8-bbd7383dd969

I can reproduce on my laptop:
```
$ west build -b sam_v71_xult zephyr/samples/shields/x_nucleo_53l0a1/
[...]
/opt/zephyr-sdk-13/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `main':
/home/titou/Code/midizephyr/zephyr/samples/shields/x_nucleo_53l0a1/src/main.c:149: undefined reference to `__device_dts_ord_14'
/opt/zephyr-sdk-13/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/drivers/sensor/vl53l0x/libdrivers__sensor__vl53l0x.a(vl53l0x.c.obj):(.data.vl53l0x_0_driver+0x8): undefined reference to `__device_dts_ord_105'
/opt/zephyr-sdk-13/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/drivers/sensor/vl53l0x/libdrivers__sensor__vl53l0x.a(vl53l0x.c.obj):(.data.vl53l0x_1_driver+0x8): undefined reference to `__device_dts_ord_107'
/opt/zephyr-sdk-13/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/drivers/sensor/vl53l0x/libdrivers__sensor__vl53l0x.a(vl53l0x.c.obj):(.data.vl53l0x_2_driver+0x8): undefined reference to `__device_dts_ord_107'
collect2: error: ld returned 1 exit status
```

This is caused by the references like `<&expander1 15 0>`, from the expansion board overlay, not being found. And indeed:
```
$ grep GPIO build/zephyr/.config
# CONFIG_GPIO is not set
CONFIG_GPIO_STMPE1600=y
# CONFIG_I2C_GPIO is not set
```

The GPIO expanders are never instanciated, because the GPIO subsystem is disabled, hence the missing references.

Therefore, I believe that adding this default is required. What do you think ?

erwango commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -246,9 +246,9 @@ static int vl53l0x_sample_fetch(const struct device *dev,
-			|| chan == SENSOR_CHAN_PROX);
+	__ASSERT_NO_MSG((chan == SENSOR_CHAN_ALL)
+			|| (chan == SENSOR_CHAN_DISTANCE)
+			|| (chan == SENSOR_CHAN_PROX));
```
Please don't add new commits to include suggestions/fixes, but integrate changes directly into impacted commits.

titouanc replied:
Thanks.
I have again rebased on top of the latest origin/main, and squashed this commit with the previous one.

erwango reviewed (changes requested):
Otherwise LGTM

titouanc commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -278,6 +234,119 @@ static int vl53l0x_init(const struct device *dev)
+	} else {
+		val->val1 = drv_data->RangingMeasurementData.RangeMilliMeter / 1000;
+		val->val2 = (drv_data->RangingMeasurementData.RangeMilliMeter % 1000) * 1000;
+	}
```
What would be the meaning of `sensor_channel_get(sensor, SENSOR_CHAN_ALL)` ?
Maybe you would have an example in another sensor driver ?

MaureenHelm replied:
It means get all of the channels supported by the sensor. See any other sensor driver

titouanc replied:
As far as I have looked in other sensor drivers, I see that `SENSOR_CHAN_ALL` is handled in `sensor_sample_fetch()` _(which is also the case in this vl53l0x driver)_ but never in `sensor_channel_get()`. Actually, the other sensor drivers return `-ENOTSUP` or `-EINVAL` when called with `SENSOR_CHAN_ALL`.

Moreover, I still do not understand how we could return multiple channels in a single `struct sensor_value`.

Maybe you expect me to replace the assertion
```c
__ASSERT_NO_MSG(chan == SENSOR_CHAN_DISTANCE || chan == SENSOR_CHAN_PROX);
```
with a proper case handling like this ?
```c
switch (chan){
    case SENSOR_CHAN_DISTANCE: ...
    case SENSOR_CHAN_PROX: ...
    default: return -ENOTSUP;
}
```

titouanc replied:
@MaureenHelm I applied my last suggestion in the 4th commit `drivers: sensor: vl53l0x: add support to reprogram I2C address`, where this function is moved, and rebased once again on top of `origin/main`

titouanc commented:
Hello @MaureenHelm and @erwango !

I pushed some changes a few weeks ago to address the split of the driver between runtime and config data. There is still one open question from @MaureenHelm (see above) for which I would like some clarifications. Otherwise, are there still some blockers that prevent to merge this PR ?

MaureenHelm commented on samples/shields/x_nucleo_53l0a1/src/display_7seg.c:
```diff
@@ -0,0 +1,165 @@
+		return -1;
+	}
+
+	expander2 = device_get_binding("EXPANDER2");
```
Use `DEVICE_DT_GET()` instead

MaureenHelm commented on samples/shields/x_nucleo_53l0a1/src/main.c:
```diff
@@ -0,0 +1,152 @@
+	gpio_add_callback(button.port, &button_cb_data);
+
+	for (i = 0; i < N_SENSORS; i++) {
+		sensors[i].dev = device_get_binding(sensors[i].name);
```
Use `DEVICE_DT_GET()` instead

avisconti commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -280,9 +284,20 @@ static int vl53l0x_init(const struct device *dev)
+#define VL53L0X_INIT(inst)						 \
+	static struct vl53l0x_config vl53l0x_##inst##_config = {	 \
+		.i2c = DEVICE_DT_GET(DT_INST_BUS(inst)),		 \
+		.dt_addr = DT_INST_REG_ADDR(inst),			 \
```
You can use i2c_dt_spec. which incorporates i2c bus device and addr, and use I2C_DT_SPEC_INST_GET() to get them. So in .h you can define 

i2c_dt_spec i2c;

And here:
.i2c = I2C_DT_SPEC_INST_GET(inst);

avisconti replied:
why did you mark it as resolved?

titouanc replied:
Good catch, thank you @avisconti 
I messed up while cherry-picking changes back into the PR commits

avisconti replied:
Yes, but not sure my comments is clear. Is it?

titouanc replied:
Yes it is, see https://github.com/zephyrproject-rtos/zephyr/pull/38315/commits/b26d4f83ec8766a0532c0e6df69672e566f226c0#diff-d58092bb6d79c1da0579e21fb5d1eed6322aa15905562ba99c8e7d0927be914bL43-L44 and https://github.com/zephyrproject-rtos/zephyr/pull/38315/commits/b26d4f83ec8766a0532c0e6df69672e566f226c0#diff-d58092bb6d79c1da0579e21fb5d1eed6322aa15905562ba99c8e7d0927be914bL289-L290

Is that what you expected ?

MaureenHelm replied:
The code change is correct, but it's in the wrong commit. If you review the commits one by one, you can see that this first gets added as `.i2c = DEVICE_DT_GET(DT_INST_BUS(inst))` in commit `drivers: sensor: vl53l0x: add driver support for multiple sensors`. Then it gets corrected to `.i2c = I2C_DT_SPEC_INST_GET(inst),` in commit `drivers: sensor: vl53l0x: add support to reprogram I2C address`.

avisconti replied:
OK, I see. I was looking at  "drivers: sensor: vl53l0x: add driver support for multiple sensors" commit only, where I think the change should be done. Can you move it in this commit? I think it behaves more to this one. Thanks and sorry for confusion.

avisconti commented:
@richardbarlow Thanks for reporting it. I'll fix it asap.

erwango commented:
@MaureenHelm Can you have a new look to this one?

titouanc commented:
Hello all,

I have once again rebased on top of the latest `main` to resolve conflicts. @MaureenHelm do you still have change requests ?
This PR is stalling for quite some time now, is there any chance this could make it into 3.0.0 ?

MaureenHelm commented:
Please address the twister and documentation failures. They do look related to these changes
